src/doc/trpl/guessing-game.md

```diff
@@ -0,0 +1,991 @@
+crates you have, and what versions you require. In this case, we’ve used `*`,
+which means that we’ll use the latest version of `rand`. Cargo understands
+[Semantic Versioning][semver], which is a standard for writing version
+numbers. If we wanted a specific version or range of versions, we could be
+more specific here. [Cargo’s documentation][cargodoc] contains more details.
+
+[semver]: http://semver.org
+[cargodoc]: http://doc.crates.io/crates-io.html
+
+Now, without changing any of our code, let’s build our project:
+
+```bash
+$ cargo build
+    Updating registry `https://github.com/rust-lang/crates.io-index`
+ Downloading rand v0.3.8
+ Downloading libc v0.1.6
+   Compiling libc v0.1.6
+   Compiling rand v0.3.8
+   Compiling guessing_game v0.1.0 (file:///home/you/projects/guessing_game)
+```
+
+(You may see different versions, of course.)
+
+Lots of new output! Now that we have an external dependency, Cargo fetches the
+latest versions of everything from the registry, which is a copy of data from
+[Crates.io][cratesio]. Crates.io is where people in the Rust ecosystem
+post their open source Rust projects for others to use.
+
+[cratesio]: https://crates.io
+
+After updating the registry, Cargo checks our `[dependencies]` and downloads
+any we don’t have yet. In this case, while we only said we wanted to depend on
+`rand`, we’ve also grabbed a copy of `libc`. This is because `rand` depends on
+`libc` to work. After downloading them, it compiles them, and then compiles
+our project.
+
+If we run `cargo build` again, we’ll get different output:
+
+```bash
+$ cargo build
+```
+
+That’s right, no output! Cargo knows that our project has been built, and that
+all of its dependencies are built, and so there’s no reason to do all that
+stuff. With nothing to do, it simply exits. If we open up `src/main.rs` again,
+make a trivial change, and then save it again, we’ll just see one line:
+
+```bash
+$ cargo build
+   Compiling guessing_game v0.1.0 (file:///home/you/projects/guessing_game)
+```
+
+So, we told Cargo we wanted any version of `rand`, and so it fetched the
+latest version at the time this was written, `v0.3.8`. But what happens
+when next week, version `v0.4.0` comes out, which changes something with
+`rand`, and it includes a breaking change? After all, a `v0.y.z` version
+in SemVer can change every release.
+
+The answer to this problem is the `Cargo.lock` file you’ll now find in your
+project directory. When you build your project for the first time, Cargo
+figures out all of the versions that fit your criteria, and then writes them
+to the `Cargo.lock` file. When you build your project in the future, Cargo
+will see that the `Cargo.lock` file exists, and then use that specific version
+rather than do all the work of figuring out versions again. This lets you
+have a repeatable build automatically.
+
+What about when we _do_ want to use `v0.4.0`? Cargo has another command,
```

Comment:
Cargo takes `0.x` to be the major releases for `0.` versions, otherwise there wouldn't be a way to make incompatible releases other than going straight to `1.0`. ([docs](http://doc.rust-lang.org/semver/semver/index.html#requirements))

In fact, strictly speaking even this is more strict than semver really is:

> Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.

I.e. `0.3.1` is "meant" to be not compatible with `0.3.2`, although I prefer our scheme.